### PR TITLE
Fix bugs found by static analysis

### PR DESCRIPTION
**Pull request by SiebelsTim**
this include:
- Comparing float against equality
- bad locking variables
- relying on default charsets
- compareTo, equal, hashCode discrepancies
- using platform specific newline in printf

I also included travis to run these and reports errors.
Of course this will only be practical if there are no false positives.
Let's see how good this works.

### CODE REVIEW

**Comment by SiebelsTim** [comment]
Note: I changed the `3.14159f`in tests because FindBugs assumed I was using a bad estimate of Pi. This was not an error that would stop travis from accepting (i.e. priority below high). 
I changed it anyways.

File: src/asmcup/genetics/Gene.java [hunk 1]
```diff
@@ -10,16 +12,27 @@ public Gene(byte[] dna, int gen, float score) {
...
 		this.score = score;
 	}
-	
+
+	@Override
 	public int compareTo(Gene other) {
 		float d = score - other.score;
 		
-		if (d == 0) {
+		if (Math.abs(d) < 0.00000001f) {
```
Comment (MLanghof) [inline src/asmcup/genetics/Gene.java]: Static analysis can say what it wants, but the check here is precisely meant to be "if this equals float zero (positive or negative)". This is (aside from NaN, which isn't handled either way) the only case where `d < 0` and `d > 0` are both false (and consequently where `a > b` and `b > a` could simultaneously be true, which is what a sort algorithm will freak out about).

I disagree with this change.

File: src/asmcup/genetics/Gene.java [hunk 2]
```diff
@@ -10,16 +12,27 @@ public Gene(byte[] dna, int gen, float score) {
...
 	}
+
+	@Override
+	public boolean equals(Object obj) {
+		return obj instanceof Gene && compareTo((Gene) obj) == 0;
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(score);
```
Comment (MLanghof) [inline src/asmcup/genetics/Gene.java]: There will be _tons_ of genes with score zero in the GA. If anybody decides to do use any construct that assumes hashes to almost always be different, that construct will be bogged down by hash collision handling.

File: src/asmcup/runtime/World.java [hunk 3]
```diff
@@ -8,9 +8,11 @@
 	protected final HashMap<Integer, byte[]> tileData;
 	protected final int seed;
 	protected int frame;
+
+	private static final Random random = new Random();
 	
 	public World() {
-		this(new Random().nextInt());
+		this(random.nextInt());
```
Comment (MLanghof) [inline src/asmcup/runtime/World.java:15]: How about `Math.random()`?

**Comment by MLanghof** [comment]
Oops, I didn't realize that would result in this much spam in here. I'll just do the review thing next time.

File: src/asmcup/runtime/World.java (hunk 3 above)
Comment (SiebelsTim) [reply src/asmcup/runtime/World.java:15]: Math.random has a crooked distribution for ints. Shouldn't matter that much, but doesn't hurt either

File: src/asmcup/genetics/Gene.java (hunk 1 above)
Comment (SiebelsTim) [reply src/asmcup/genetics/Gene.java]: I still don't understand how you expect that if to evaluate to true when comparing for binary equivalence

File: src/asmcup/genetics/Gene.java (hunk 2 above)
Comment (SiebelsTim) [reply src/asmcup/genetics/Gene.java]: Which would happen with the current implementation anyways?

File: src/asmcup/genetics/Gene.java (hunk 1 above)
Comment (MLanghof) [reply src/asmcup/genetics/Gene.java]: ...floats can be binary equivalent. If you calculate `a-a`, you will (barring NaN etc.) always get zero.

Say we have two genes, `g1` and `g2` that, for some reason, have identical DNA (but were created separately from each other, which is not unlikely). Obviously, they will have the same score `z`. `g1.compareTo(g2)` will compute `z-z` and thereby get a `d` of precisely 0. The floats are (binary) equivalent, hence `compareTo()` must return 0.

In all other cases where `d` is non-zero, one of the floats was bigger than the other (this should be trivial to see). Hence, `compareTo()` should _not_ return zero.

File: src/asmcup/genetics/Gene.java (hunk 2 above)
Comment (MLanghof) [reply src/asmcup/genetics/Gene.java]: Not quite. The default behavior (if I'm not mistaken) is that the object pointer is hashed (or something along those lines, point being that different [in the sense of "not the exact same instance"] objects usually have different hash codes and don't equal each other).

This means that the current implementation is [not consistent with equals](http://stackoverflow.com/questions/12123960/what-does-comparison-being-consistent-with-equals-mean-what-can-possibly-happe), but doesn't bog down e.g. a hash map.

The implication of this (to my understanding) is that (currently) asking a hash map for a `new Gene(score = z)` will return nothing, even though `new Gene(score = z).compareTo(g1)` would return true.

That's not a problem, because it's important to note that our genes must be sorted based on their score, but having the same score does not imply they are equal.

File: src/asmcup/genetics/Gene.java (hunk 1 above)
Comment (SiebelsTim) [reply src/asmcup/genetics/Gene.java]: Of course floats can be binary equivalent. Relying on it to be out of a calculation is bad.

I blindly assumed `score` was something the Gene will develop, which will increase/decrease somehow. 

You were trying to compare if the DNA is equal, in a faster way than comparing arrays?

File: src/asmcup/genetics/Gene.java (hunk 2 above)
Comment (SiebelsTim) [reply src/asmcup/genetics/Gene.java]: Okay, I think I misunderstood `score`.

File: src/asmcup/genetics/Gene.java (hunk 1 above)
Comment (MLanghof) [reply src/asmcup/genetics/Gene.java]: > Of course floats can be binary equivalent. Relying on it to be out of a calculation is bad.
- We assume those floats are neither `NaN` nor `+/-Infinity`.
- Then, if the floats `a` and `b` have the **same value**, they are binary equivalent. Their difference is therefore precisely `+0` (there are only two zero values in IEEE floats, `-0` and `+0`, they also always behave as equal under numeric comparisons, so not like it matters). This is the condition being checked, nothing else.
- If the floats have **different values**, their difference `d` _cannot be zero_. Hence, either `d < 0` xor `d > 0` is true. Therefore, `a.compareTo(b) == -(b.compareTo(a)) != 0`, which is required for the sorting algorithm in question.

We're not really relying on a particular float taking a particular binary representation. We're merely guarding against it.

> You were trying to compare if the DNA is equal

Where did I state that? All we're trying to do is introduce an ordering to a set of genes. That ordering only cares about the score (because we want to sort by it). The question "when are two genes equal?" hasn't been a concern so far. For what we care about, they are only equal if they are literally the same object instance.

File: src/asmcup/genetics/Gene.java (hunk 2 above)
Comment (MLanghof) [reply src/asmcup/genetics/Gene.java]: Just for the record, every member of `Gene` is `final`.

File: src/asmcup/genetics/Gene.java (hunk 1 above)
Comment (SiebelsTim) [reply src/asmcup/genetics/Gene.java]: Excuse me, but I still don't understand it.

You are relying on both floats to be equal.

Let's say you will increase score of Gene A by `0.7f` and score of Gene B by `0.37f` and then `0.33f`.
They will not compare to be equal (`0.37f + 0.33f != 0.7f`).

> For what we care about, they are only equal if they are literally the same object instance.
> But you don't check for it. You check for score having the same binary representation.

I understand that you are creating an ordering. But it's arbitrary. 

After all I don't think it's worth fighting over this, as I don't seem to understand it anyway.
Additionally, I haven't looked into Genes and might be missing the crucial point.

I've already reverted this hours ago.

Comment (MLanghof) [reply src/asmcup/genetics/Gene.java]: (Note: I'm not arguing because I'm trying to "fight" over this, just trying to make the intentions clear.)

> Let's say you will increase score of Gene A by 0.7f and score of Gene B by 0.37f and then 0.33f.

This cannot happen because the scores are final. But even if it were possible:

> I understand that you are creating an ordering. But it's arbitrary.

...not more arbitrary than the float calculations.

All we're trying to do is order the genes by score. If, due to whatever calculations happened before, two scores that should be equal are not equal, then so be it. For what it's worth to the comparison of genes, one of the two is bigger, which is all that piece of code has to determine.

In other words, the goal of the code in `compareTo()` is not primarily to determine equality. In fact, in an initial version there was no check for equality, so in the situation that `!(a > b)` and `!(b > a)` both are true (which means `a == b`), `g1.compareTo(g2)` and `g2.compareTo(g1)` both returned the same (nonzero) number, implying `a > b` at the same time as `b > a`, which every sorting algorithm will (and did) throw a fit over. That's why a check for (binary) equality was introduced, just to weed out that single edge case (which is only present for `a == b <=> a-b == 0 <=> (!(a > b) and !(b > a))`).

**Comment by MLanghof** [comment]
Alright, no further objections!

File: src/asmcup/genetics/Gene.java (hunk 1 above)
Comment (SiebelsTim) [reply src/asmcup/genetics/Gene.java]: > (Note: I'm not arguing because I'm trying to "fight" over this, just trying to make the intentions clear.)

Yeah, I know. I just thought this will get out of hand and you are probably right considering that I am not familiar with the Gene codebase. No hurt feelings!

> This cannot happen because the scores are final. 

This is what I was missing. :x 

---

The last paragraph clicked. Funny that I now finally understand as soon as I thought I'll stop spending my time on this particular issue :D.

File: src/asmcup/runtime/World.java (hunk 3 above)
Comment (krisives) [reply src/asmcup/runtime/World.java:15]: @SiebelsTim Math.random() is implemented in Java as having a global Random object and calling nextDouble() on it basically: http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b27/java/lang/Math.java#Math.random%28%29

**Comment by krisives** [comment]
Thank you Tim! In theory `getWorld()` should be synchornized too.

Also note that something like this in Java:

```
public void foo() {
  synchronized (this) {
    // do things
  }
}
```

is the same as:

```
public synchronized void foo() {
```

**Comment by SiebelsTim** [comment]
Yep. However there's still a draw() outside of the synchronized block. 
The error was that the world variable would be set inside the block. Effectively allowing for two threads to enter the synchronized block.
